### PR TITLE
Feature/median module

### PR DESCRIPTION
**Pull request by AmadiMichael**
This PR introduces these changes and modifications:

- changes dev framework to foundry which has proven to be better especially in test capabilities compared to hardhat.
- fixes the issues (accepts only lower order of s in ecrecover, using eip191 msg hash and some gas optimizations).
- adds unit tests for all functions of `Median.sol`. (fuzz and invariants to follow)
- adds deployment script for local and public chain deployment (see readme)

More changes
- use bytes signature not separated
- remove sort function and perform it offchain while confirming correct sorting onchain
- can now test with relayer client and foundry script on local machine

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,66 @@
...
+### Anvil
+
+```shell
+$ anvil
+```
+
+### Deploy
+
+```shell
+$ forge script script/Counter.s.sol:CounterScript --rpc-url <your_rpc_url> --private-key <your_private_key>
```
Comment (njokuScript) [inline README.md]: make this specific to our deployment script.

Can you write a script for deploying the contract?

**Review by njokuScript — COMMENTED** [review]
It generally looks good!

One last thing will be to add a deployment script for the `median.sol` contract

File: README.md (hunk 1 above)
Comment (AmadiMichael) [reply README.md]: Done

File: src/median.sol [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        uint256 _lastTimestamp = lastTimestamp;
+        uint256 _trackedPrice;
+
+        for (uint256 i; i < _prices.length; ++i) {
+            if (_timestamps[i] <= _lastTimestamp) revert InvalidTimestamp();
+            if (_prices[i] < _trackedPrice) revert PricesNotOrdered();
+
+            address signer = recover(_prices[i], _timestamps[i], currencyPair, _signatures[i]);
+
+            // ecdsa lib already reverts with error `ECDSAInvalidSignature()` if signer == address(0)
```
Comment (iamnotstatic) [inline src/median.sol:93]: Make sense, read somewhere that In some cases ecrecover can return a random address instead of 0 for an invalid signature.

Comment (AmadiMichael) [reply src/median.sol:93]: Yeah most times it always recovers to an address, it only returns address(0) if r is not a valid x axis point on the curve, if v is not 27 or 28 and if the public key is the identity element of the curve. If not it'll recover to a valid public key with a valid address. 

That's why there's another check to ensure signer is msg.sender (even though the open zeppelin Lib reverts if it returns address 0)

**Pull request closed without merging**